### PR TITLE
Improved macros

### PR DESCRIPTION
**Pull request by delight-aug**
Hi! This time it's nothing radical. These macros have the following improvements:
- Can be used without arguments
- In `dev` build, panics with different messages for situations when there are less/more arguments than format directives
- Can have ~~one~~ `{n}` format directive for including number used as `n` in `ngettext`

Should be tested with both `cargo test` and `cargo test --release`. Tests for situations when there are less/more arguments than format directives do not test if panics happen and in which build profile.

TODO:
- [x] Document `{n}` functionality
- [x] Improve tests for "less/more arguments" panics

### CODE REVIEW

**Comment by delight-aug** [comment]
Also, I couldn't get crate to build in release profile, for some reason, so I ran release tests as regular ones by commenting out debug_asserts, to at least ensure Ok()=>asserts are correct.

**Review by Minoru — CHANGES_REQUESTED** [review]
Drafts are, ostensibly, not meant to be reviewed, but it looks like you want a review, so here goes :)

> Can be used without arguments

That's a good idea, thanks!

> Move needed tests to be integration tests
> Introduce '{{}}' escaping?

I think these are separate tasks. Please open issues for them, so we don't forget!

> Also, I couldn't get crate to build in release profile — gettext-sys was building for around 5 minutes and errored.

Errored how? Please open an issue, and let's investigate.

Finally, please rebase this PR onto the latest commit from `master`.

**Comment by delight-aug** [comment]
> Drafts are, ostensibly, not meant to be reviewed, but it looks like you want a review, so here goes :)

I didn't know, thanks! I'll resolve all the other notices, but I might not be very active for few days, don't worry about that.

File: gettext-rs/src/macros.rs [hunk 1]
```diff
@@ -71,177 +120,692 @@ macro_rules! dcgettext {
...
 }
 
 #[cfg(test)]
 mod test {
     use crate::*;
+    use std::panic::catch_unwind;
 
     #[test]
-    fn test_gettext_macro() {
+    fn no_formatting() {
```
[the remark this answers is not in the document]
Comment (delight-aug) [reply gettext-rs/src/macros.rs]: No, it was just my attempt at organizing tests.

**Review by Minoru — CHANGES_REQUESTED** [review]
Excellent work! There are just a few remaining issues with the test suite. Please see my comments below, and also run `cargo fmt` to fix the one failing CI task.

Also, you merged `master` instead of rebasing onto it. This doesn't matter for this PR because I'll squash it before merge, but please be more attentive next time ;)

File: gettext-rs/src/macros.rs [hunk 2]
```diff
@@ -57,191 +90,890 @@ macro_rules! dgettext {
...
+    #[cfg_attr(debug_assertions, should_panic="There are fewer arguments than format directives")]
+    fn gettext_fewer_arguments_than_parameters() {
+        setlocale(LocaleCategory::LcAll, "en_US.UTF-8");
         bindtextdomain("hellorust", "/usr/local/share/locale").unwrap();
         textdomain("hellorust").unwrap();
 
         assert_eq!(
-            "Multiple Worlds!",
-            npgettext!("context", "Singular {}!", "Multiple {}!", 2, "Worlds")
+            gettext!("Hello, {}! {}", "World"),
```
[the remark this answers is not in the document]
Comment (delight-aug) [reply gettext-rs/src/macros.rs:491]: We're expecting a panic when launching with dev profile, but making assertion with release profile.

**Comment by delight-aug** [comment]
> and also run cargo fmt to fix the one failing CI task.

I already checked it, but it suggests very questionable changes:
```
-        assert_eq!(
-            dgettext!("hellorust",
-                "Hello, World!"),
-            "Hello, World!"
-        );
+        assert_eq!(dgettext!("hellorust", "Hello, World!"), "Hello, World!");
```
Writing this assertion in one line only decreases readability. While the next one doesn't really change anything:
```
-        Some((pre, suf)) => { format!("{}{}{}", pre, arg, suf) }
+        Some((pre, suf)) => {
+            format!("{}{}{}", pre, arg, suf)
+        }
```
Other proposed changes are similar. Considering the `cargo fmt` check is not required, I'd rather skip it.

File: gettext-rs/src/macros.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Minoru) [reply gettext-rs/src/macros.rs:491]: Oh, right >_< My bad!

**Comment by Minoru** [comment]
I agree that some of the choices that `cargo fmt` made are suboptimal, but its benefits far outweigh these "costs". It's a blessing that one can fix all the style issues with a single command. It's even more of a blessing that many other Rust projects use the same style too. It doesn't mean everyone *likes* the style, or that the style always leads to the best possible look, but the fact that it's *consistent* is far more important than personal taste or local inconveniences.

If you feel strongly about this, you can nudge `cargo fmt` to produce what you want. You can remove curly braces if you want that block on a single line. You can make the strings longer so they don't fit onto a single line. It's extra work, but at least you don't rob future developers from the convenience of `cargo fmt`.

I think it's an error that this check is not required. If we skip it once, it will keep failing for all future PRs, which makes the check useless. @Koka, can you please go to _Settings_ → _Branches_, in _Branch protection rules_ click _Edit_ next to "master", and make the "Code formatting" check required? I can't do this myself because I don't have access to the *Settings* tab.

**Review by Minoru — APPROVED** [review]
The code is fine, all that's left to do is reach a consensus on formatting.

**Comment by Koka** [comment]
I'd be glad to enable such a protection but apparently there's no such check there

**Comment by Minoru** [comment]
@Koka the interface is not great there: they show you the checks that are *already* required, and if you want to add more, you have to type into the search field and choose from the popup that appears.

**Comment by Minoru** [comment]
Thanks @delight-aug! I'm happy to see all your hard work finally come to fruition :tada: I expect you'd tackle #72 next?

@Koka, I merge this, but please take another look at those required checks, it'd be really nice to make the formatting required.

**Comment by Minoru** [comment]
Oh, I forgot to squash >_< Now we have to live with a couple kinks in our history graph, sorry about that everyone!

**Comment by delight-aug** [comment]
Thank you, @Minoru, for helping me out with my first programming contribution! I couldn't do that, if you weren't so open!

> I expect you'd tackle #72 next?

I want to work on #71 first, so I have proper testing for working on #72.

**Comment by Koka** [comment]
@Minoru I tried everything and yet there's no such option in branch protection https://monosnap.com/file/QGVPGH53txZ5a06tVYL3VnRocVBO9c

**Comment by Minoru** [comment]
@Koka, at 00:11 when you typed "rust", "rustfmt" appeared as the very last item in the list, half-hidden because the damn list is supposed to be scrolled but doesn't show the scroll bar. I only had to re-watch the video two times to notice that :sweat_smile: What a terrible UI they have!

**Comment by Koka** [comment]
@Minoru Alright! After third attempt I've seemingly done it 😳 Thanks for your guidance!

**Pull request merged**